internal/restorer/preallocate_darwin.go

```diff
@@ -0,0 +1,33 @@
+package restorer
+
+import (
+	"os"
+	"runtime"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+func preallocateFile(wr *os.File, size int64) error {
+	// try contiguous first
+	fst := unix.Fstore_t{
```

Comment:
looks plausible, but I am not familiar with unit filesystem API to review this properly. ditto for other planforms.

Reply:
I've tested the preallocate implementations for Linux and macOS manually:

On macOS I've used the code snippet from https://apple.stackexchange.com/a/37844 to count the file fragments. For the preallocated files the number of fragments dropped from several to one (and a few outliers with up to three fragments, I think).

On Linux `filefrag` showed that the number of fragments went down to 1 or 2.

The macOS implementation is inspired by https://searchfox.org/mozilla-central/source/xpcom/glue/FileUtils.cpp#67 and https://github.com/etcd-io/etcd/blob/master/pkg/fileutil/preallocate_darwin.go#L35 . The main difference is that my parameter selection first tries to allocate the whole file as one continuous chunk and if that fails then just asks for the whole file to be preallocated.

Reply:
Do you think "seems to work" is good enough for the `preallocate` implementations? I've also just compared the implementation with that of etcd and it looks roughly similar. For Linux etcd falls back to `ftruncate` when the `fallocate` call has failed. I'm not really sure in which cases this would help.

Reply:
yes, this is good enough for me.